crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/bloom.rs

```diff
@@ -0,0 +1,174 @@
+            bloom.set(&key);
+        }
+        Ok(Some(bloom))
+    }
+
+    /// Generate a bloom filter of all ops.
+    /// - Ops are only included if they are within the common arc set.
+    /// - The bloom is `KitsuneOpHah`.
+    /// - Ops are only included from local agents that are within the common arc set.
+    /// - Only create the filter if there are any ops matching the above criteria.
+    /// No empty bloom filters.
+    /// - Bloom has a 1% chance of false positive (which will lead to agents not being sent back).
+    /// - Expect this function to complete in an average of 10 ms and worst case 100 ms.
+    pub(super) async fn generate_ops_blooms_for_time_window(
+        &self,
+        local_agents: &HashSet<Arc<KitsuneAgent>>,
+        common_arc_set: &Arc<DhtArcSet>,
+        mut search_time_window: Range<u64>,
+    ) -> KitsuneResult<Vec<TimedBloomFilter>> {
+        let mut results = Vec::new();
+        loop {
+            // Get the agents withing the arc set and filter by local.
+            let local_agents_within_arc_set: Vec<_> =
+                store::agents_within_arcset(&self.evt_sender, &self.space, common_arc_set.clone())
+                    .await?
+                    .into_iter()
+                    .filter(|(a, _)| local_agents.contains(a))
+                    .collect();
+
+            // Get the op hashes which fit within the common arc set from these local agents.
+            let result = store::all_op_hashes_within_arcset(
+                &self.evt_sender,
+                &self.space,
+                local_agents_within_arc_set.as_slice(),
+                common_arc_set,
+                search_time_window.clone(),
+                Self::UPPER_HASHES_BOUND,
+                true,
+            )
+            .await?;
+
+            // If there are none then don't create a bloom.
+            let (ops_within_common_arc, found_time_window) = match result {
+                Some(r) => r,
+                None => break,
+            };
+
+            let num_found = ops_within_common_arc.len();
+
+            // Create the bloom from the op hashes.
+            let mut bloom =
+                bloomfilter::Bloom::new_for_fp_rate(ops_within_common_arc.len(), Self::TGT_FP);
+
+            for hash in ops_within_common_arc {
+                bloom.set(&Arc::new(MetaOpKey::Op(hash)));
+            }
+
+            // If we found the maximum number of ops we can then
+            // there might still be more ops in the search window.
+            if num_found >= Self::UPPER_HASHES_BOUND
+                && found_time_window.end < search_time_window.end
+            {
+                let bloom = TimedBloomFilter {
+                    bloom,
+                    time: search_time_window.start..found_time_window.end,
+                };
+                // Adjust the search window to search the remaining time window.
+                search_time_window = found_time_window.end..search_time_window.end;
```

Comment:
I looked to make sure this exclusive endpoint would include the last hash. It does, so I added a comment.

```suggestion
                let bloom = TimedBloomFilter {
                    bloom,
                    // the `time_window.end` is exclusive, but has already been incremented
                    // to account for this
                    time: search_time_window.start..found_time_window.end,
                };
                // Adjust the search window to search the remaining time window.
                search_time_window = found_time_window.end..search_time_window.end;
```